hcloud/cloud.go

```diff
@@ -201,6 +215,23 @@ func loadBalancerDefaultsFromEnv() (hcops.LoadBalancerDefaults, bool, error) {
 	return defaults, disablePrivateIngress, nil
 }
 
+// serverIsAttachedToNetwork checks if the server where the master is running on is attached to the configured private network
+// We use this measurement to protect users against some parts of misconfiguration, like configuring a master in a not attached
+// network.
+func serverIsAttachedToNetwork(networkId int) (bool, error) {
+	const op = "serverIsAttachedToNetwork"
+	resp, err := http.Get("http://169.254.169.254/hetzner/v1/metadata/private-networks")
+	if err != nil {
+		return false, fmt.Errorf("%s: %s", op, err)
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return false, fmt.Errorf("%s: %s", op, err)
+	}
+	return strings.Contains(string(body), fmt.Sprintf("network_id: %d", networkId)), nil
```

Comment:
```suggestion
	return strings.Contains(string(body), fmt.Sprintf("network_id: %d\n", networkId)), nil
```

I suggest adding at least the newline to the search string, so we can't accidentally match a network with a longer ID and the same prefix. Additionally, you could check for whitespace before `network_id`, so we don't accidentally match another field, in case we ever add one (e.g. `public_network_id`).

If you apply this, you'll want to check whether we actually have single newlines (`\n`) in the metadata.

Or, you could use the `gopkg.in/yaml.v2` package that's already in the CCM's dependencies to do this properly 😄